File: src/Audentity/Tracking/Trace.cs
using Microsoft.EntityFrameworkCore;

namespace Audentity.Tracking;

public record Trace(string Name)
{
    public IEnumerable<Property> Properties { get; init; } = Enumerable.Empty<Property>();
    public IEnumerable<Reference> References { get; init; } = Enumerable.Empty<Reference>();
    public EntityState State { get; internal init; } = EntityState.Unchanged;
    public bool IsOwned { get; init; }
}

File: src/Audentity/Trace.fs
namespace Audentity

open System
open System.Collections.Generic
open Microsoft.EntityFrameworkCore

type Link = { Name: string; Value: string option }

type Reference =
    { Name: string
      Type: Type
      Links: IReadOnlyCollection<Link> }

type Property =
    { Name: string
      CurrentValue: string option
      OriginalValue: string option
      IsPrimaryKey: bool }

type Trace =
    { Type: Type
      State: EntityState
      Properties: IReadOnlyCollection<Property>
      References: IReadOnlyCollection<Reference>
      IsOwned: bool }
